### ISSUE TITLE
max_p, max_d, max_q parameters set by user

### ISSUE THREAD

**Issue by mdeligian** [alkaline-ml/pmdarima#405]
**Describe the bug**
The function "auto_arima" has hard-coded (fixed-values) for parameters max_p, max_d, max_q, which are max_p = 5, max_d = 2, max_q = 5.

**To Reproduce**  
Run any auto_arima model with the aforementioned parameters greater than 5.

```python
arima_model =  auto_arima(data, max_p = 10, max_d = 10, max_q  = 10)
```

**Versions**
System:
    python: 3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)]
executable: C:\Users\MD\python.exe
   machine: Windows-10-10.0.18362-SP0

Python dependencies:
        pip: 20.1.1
 setuptools: 47.3.1.post20200622
    sklearn: 0.23.1
statsmodels: 0.11.1
      numpy: 1.18.5
      scipy: 1.5.0
     Cython: 0.29.17
     pandas: 1.0.5
     joblib: 0.15.1
   pmdarima: 1.8.0
-->

**Expected behavior**
The values of the parameters should be provided by the user.

**Actual behavior**
The values of the parameters are fixed.

**Additional context**

**Comment by NowanIlfideme** [comment]
Not a pmdarima developer, but... are there any real-world use cases with a true level of integration (d param) greater than 2? The AR/MA (p, q) parameters are a different question, of course.

**Comment by mdeligian** [comment]
@NowanIlfideme The purpose of this issue was not to review the statistical properties / usage of ARIMA modelling techniques but to mention that probably there is a bug in the implementation of "auto_arima", i.e. probably there is a misalignment in between what the code should suppose to do and what it does.

**Comment by tgsmith61591** [comment]
I don't think this is a bug. If you run the following, you'll see an exhaustive grid search all the way up to your maximum values:

```python
In [1]: import pmdarima as pm
In [2]: y = pm.datasets.load_wineind()
In [3]: model = pm.auto_arima(y, m=12, max_p=10, max_d=10, max_q=1, start_q=1, trace=3, stepwise=False, max_order=None)
 ARIMA(0,1,0)(0,1,0)[12]             : AIC=3131.408, Time=0.03 sec
 ARIMA(0,1,0)(0,1,1)[12]             : AIC=3117.921, Time=0.34 sec
 ARIMA(0,1,0)(0,1,2)[12]             : AIC=3116.480, Time=1.33 sec
    ...
 ARIMA(10,1,1)(1,1,2)[12]             : AIC=inf, Time=10.68 sec
 ARIMA(10,1,1)(2,1,0)[12]             : AIC=3082.933, Time=9.75 sec
 ARIMA(10,1,1)(2,1,1)[12]             : AIC=3082.894, Time=11.49 sec
Near non-invertible roots for order (10, 1, 1)(2, 1, 2, 12); setting score to inf (at least one inverse root too close to the border of the unit circle: 0.995)
 ARIMA(10,1,1)(2,1,2)[12]             : AIC=inf, Time=16.77 sec

Best model:  ARIMA(0,1,1)(0,1,1)[12]
Total fit time: 553.851 seconds
```

By the way, a stationarity test is used for `d`, so it will not perform a search over that value; it evaluates it one time only.

But what I suspect you're referring to is the stepwise behavior, where it's a bit less intuitive how the model takes steps in any given direction. The stepwise algo will use your starting parameters as a base, but then it will begin from a null model, taking steps in either direction. Here is an example:

```python
In [4]: model = pm.auto_arima(y, m=12, max_p=10, start_p=7, max_q=10, start_q=8, trace=3)
Performing stepwise search to minimize aic
Near non-invertible roots for order (7, 1, 8)(1, 1, 1, 12); setting score to inf (at least one inverse root too close to the border of the unit circle: 1.000)
 ARIMA(7,1,8)(1,1,1)[12]             : AIC=inf, Time=11.28 sec
 ARIMA(0,1,0)(0,1,0)[12]             : AIC=3131.408, Time=0.09 sec
First viable model found (3131.408)
 ARIMA(1,1,0)(1,1,0)[12]             : AIC=3097.884, Time=0.28 sec
New best model found (3097.884 < 3131.408)
 ARIMA(0,1,1)(0,1,1)[12]             : AIC=3066.329, Time=0.24 sec
New best model found (3066.329 < 3097.884)
 ARIMA(0,1,1)(0,1,0)[12]             : AIC=3089.456, Time=0.09 sec
 ARIMA(0,1,1)(1,1,1)[12]             : AIC=3067.457, Time=0.43 sec
 ARIMA(0,1,1)(0,1,2)[12]             : AIC=3067.481, Time=0.66 sec
 ARIMA(0,1,1)(1,1,0)[12]             : AIC=3071.631, Time=0.22 sec
Near non-invertible roots for order (0, 1, 1)(1, 1, 2, 12); setting score to inf (at least one inverse root too close to the border of the unit circle: 1.000)
 ARIMA(0,1,1)(1,1,2)[12]             : AIC=inf, Time=3.99 sec
 ARIMA(0,1,0)(0,1,1)[12]             : AIC=3117.921, Time=0.32 sec
 ARIMA(1,1,1)(0,1,1)[12]             : AIC=3066.492, Time=0.54 sec
 ARIMA(0,1,2)(0,1,1)[12]             : AIC=3065.533, Time=0.24 sec
New best model found (3065.533 < 3066.329)
 ARIMA(0,1,2)(0,1,0)[12]             : AIC=3087.883, Time=0.11 sec
 ARIMA(0,1,2)(1,1,1)[12]             : AIC=3066.239, Time=0.42 sec
 ARIMA(0,1,2)(0,1,2)[12]             : AIC=3066.373, Time=0.65 sec
 ARIMA(0,1,2)(1,1,0)[12]             : AIC=3070.728, Time=0.22 sec
Near non-invertible roots for order (0, 1, 2)(1, 1, 2, 12); setting score to inf (at least one inverse root too close to the border of the unit circle: 0.998)
 ARIMA(0,1,2)(1,1,2)[12]             : AIC=inf, Time=2.40 sec
 ARIMA(1,1,2)(0,1,1)[12]             : AIC=3066.424, Time=0.54 sec
 ARIMA(0,1,3)(0,1,1)[12]             : AIC=3066.351, Time=0.51 sec
 ARIMA(1,1,3)(0,1,1)[12]             : AIC=3068.295, Time=1.02 sec
 ARIMA(0,1,2)(0,1,1)[12] intercept   : AIC=3066.787, Time=0.34 sec

Best model:  ARIMA(0,1,2)(0,1,1)[12]
Total fit time: 24.639 seconds
```

As you can see, it will start with `(7,1,8)`, but then moves values around a null model. This actually matches the behavior of R's implementation (keep in mind subtle differences in how R and python calculate the AIC will cause minor detours in model selection):

```R
> library(forecast)
> auto.arima(wineind, max.p=10, start.p=7, max.q=10, start.q=8, trace=TRUE)

 Fitting models using approximations to speed things up...

 ARIMA(7,1,8)(1,1,1)[12]                    : Inf
 ARIMA(0,1,0)(0,1,0)[12]                    : 2918.738
 ARIMA(1,1,0)(1,1,0)[12]                    : 2880.387
 ARIMA(0,1,1)(0,1,1)[12]                    : 2789.402
 ARIMA(0,1,1)(1,1,1)[12]                    : 2797.908
 ARIMA(0,1,1)(0,1,0)[12]                    : 2827.788
 ARIMA(0,1,1)(0,1,2)[12]                    : 2788.526
 ARIMA(1,1,1)(0,1,2)[12]                    : 2788.043
 ARIMA(1,1,0)(0,1,2)[12]                    : 2851.21
 ARIMA(1,1,2)(0,1,2)[12]                    : Inf
 ARIMA(0,1,0)(0,1,2)[12]                    : 2885.681
 ARIMA(2,1,2)(0,1,2)[12]                    : Inf
 ARIMA(1,1,1)(1,1,2)[12]                    : 2797.239
 ARIMA(1,1,1)(0,1,1)[12]                    : 2787.4
 ARIMA(2,1,1)(0,1,1)[12]                    : 2789.621
 ARIMA(1,1,0)(0,1,1)[12]                    : 2853.877
 ARIMA(1,1,2)(0,1,1)[12]                    : 2787.155
 ARIMA(2,1,3)(0,1,1)[12]                    : Inf
 ARIMA(1,1,2)(1,1,1)[12]                    : 2799.39
 ARIMA(1,1,2)(0,1,0)[12]                    : Inf
 ARIMA(1,1,2)(1,1,2)[12]                    : 2798.977
 ARIMA(0,1,2)(0,1,1)[12]                    : 2789.574
 ARIMA(2,1,2)(0,1,1)[12]                    : Inf
 ARIMA(1,1,3)(0,1,1)[12]                    : Inf

 Now re-fitting the best model(s) without approximations...

 ARIMA(1,1,2)(0,1,1)[12]                    : 3004.485

 Best model: ARIMA(1,1,2)(0,1,1)[12]
```

#### The bottom line

`pmdarima`'s ultimate goal is to match the R package's behavior. In this case, while the stepwise behavior may seem odd, it is doing the same thing as R which we would consider "to spec." If your goal is to do a broad sweep with very high values, then please take the first approach with `stepwise=False`. I _do_ think it may be worth explaining this behavior in a section of documentation.

Does this make sense?

**Comment by mdeligian** [comment]
@tgsmith61591 Thank you for your prompt answer. Your insights are really helpful. After some experimentation, I have to note that if the user wants to search exhaustively (stepwise=False) then "pmdarima" tries out all combinations for p and q but not for d. Please, see the example below:

import pmdarima as pm
y = pm.datasets.load_wineind()
model = pm.auto_arima(y, max_p=7, max_d=10, max_q=2,
                      start_p=1, start_q=1, 
                      trace=3, stepwise=False, max_order=None,
                      seasonal=False)
 ARIMA(0,1,0)(0,0,0)[0] intercept   : AIC=3587.776, Time=0.07 sec
 ARIMA(0,1,1)(0,0,0)[0] intercept   : AIC=3509.903, Time=0.59 sec
 ARIMA(0,1,2)(0,0,0)[0] intercept   : AIC=3506.259, Time=1.05 sec
 ARIMA(1,1,0)(0,0,0)[0] intercept   : AIC=3573.940, Time=0.12 sec
 ARIMA(1,1,1)(0,0,0)[0] intercept   : AIC=3508.854, Time=1.09 sec
 ARIMA(1,1,2)(0,0,0)[0] intercept   : AIC=3506.161, Time=1.09 sec
 ARIMA(2,1,0)(0,0,0)[0] intercept   : AIC=3540.400, Time=0.16 sec
 ARIMA(2,1,1)(0,0,0)[0] intercept   : AIC=3500.610, Time=0.51 sec
 ARIMA(2,1,2)(0,0,0)[0] intercept   : AIC=3503.442, Time=0.83 sec
 ARIMA(3,1,0)(0,0,0)[0] intercept   : AIC=3516.118, Time=0.21 sec
 ARIMA(3,1,1)(0,0,0)[0] intercept   : AIC=3502.598, Time=0.85 sec
 ARIMA(3,1,2)(0,0,0)[0] intercept   : AIC=3504.888, Time=0.48 sec
 ARIMA(4,1,0)(0,0,0)[0] intercept   : AIC=3516.442, Time=0.26 sec
 ARIMA(4,1,1)(0,0,0)[0] intercept   : AIC=3501.754, Time=0.87 sec
 ARIMA(4,1,2)(0,0,0)[0] intercept   : AIC=3502.665, Time=1.34 sec
 ARIMA(5,1,0)(0,0,0)[0] intercept   : AIC=3517.051, Time=0.28 sec
 ARIMA(5,1,1)(0,0,0)[0] intercept   : AIC=3502.515, Time=1.17 sec
 ARIMA(5,1,2)(0,0,0)[0] intercept   : AIC=3503.431, Time=1.16 sec
 ARIMA(6,1,0)(0,0,0)[0] intercept   : AIC=3504.134, Time=0.27 sec
 ARIMA(6,1,1)(0,0,0)[0] intercept   : AIC=3485.201, Time=0.72 sec
 ARIMA(6,1,2)(0,0,0)[0] intercept   : AIC=3487.257, Time=0.83 sec
 ARIMA(7,1,0)(0,0,0)[0] intercept   : AIC=3502.162, Time=0.31 sec
 ARIMA(7,1,1)(0,0,0)[0] intercept   : AIC=3486.306, Time=0.54 sec
Near non-invertible roots for order (7, 1, 2)(0, 0, 0, 0); setting score to inf (at least one inverse root too close to the border of the unit circle: 0.991)
 ARIMA(7,1,2)(0,0,0)[0] intercept   : AIC=inf, Time=2.04 sec

Best model:  ARIMA(6,1,1)(0,0,0)[0] intercept
Total fit time: 17.039 seconds

 As you refer to in your comments perhaps this happens due to an enforced stationarity test.

**Comment by tgsmith61591** [comment]
Correct, see [this block of code](https://github.com/alkaline-ml/pmdarima/blob/master/pmdarima/arima/auto.py#L560):

```python
    # determine/set the order of differencing by estimating the number of
    # orders it would take in order to make the TS stationary.
    if d is None:
        d = ndiffs(
            dx,
            test=test,
            alpha=alpha,
            max_d=max_d,
            **offset_test_args,
        )

        if d > 0 and X is not None:
            diffxreg = diff(diffxreg, differences=d, lag=1)

            # if any columns are constant, subtract one order of differencing
            if np.apply_along_axis(is_constant, arr=diffxreg, axis=0).any():
                d -= 1
```

This, again, is how R is doing it. It's good to note the `auto_arima` function is not simply intended to be a fancy wrapper of a nested for-loop; certain parameters are learned once, while others are searched over.
